### DISCUSSION TITLE
Save into file

### DISCUSSION THREAD

**Discussion by sdavidg** [Aris-t2/CustomJSforFx#100] [Q&A] [answered]
I use this code to save a string into a file:
```javascript
var ostream = Cc['@mozilla.org/network/file-output-stream;1'].createInstance(Ci.nsIFileOutputStream);
ostream.init(file, -1, -1, 0);
var converter = Cc['@mozilla.org/intl/scriptableunicodeconverter'].createInstance(Ci.nsIScriptableUnicodeConverter);
converter.charset = 'UTF-8';
var istream = converter.convertToInputStream(content);
NetUtil.asyncCopy(istream, ostream)
```
Starting with FF119, the function **converter.convertToInputStream** don't exist anymore.

I'd try 
```javascript
istream = Cc["@mozilla.org/io/string-input-stream;1"].createInstance(Ci.nsIStringInputStream);
istream.data =  content
```
but I have problems with characters like **á**, **ó**, **ñ** etc

Does anyone have any idea how to continue?

**Comment by 117649** [comment, marked as answer]
by Mozilla:
```js
const istream = Cc["@mozilla.org/io/string-input-stream;1"].createInstance(Ci.nsIStringInputStream);
istream.setUTF8Data(codeElementWrapper.value);
```

**Comment by sdavidg** [comment]
I was testing this one:
```javascript
function encode_utf8(s) {
	return unescape(encodeURIComponent(s));
}
var istream = Cc["@mozilla.org/io/string-input-stream;1"].createInstance(Ci.nsIStringInputStream);
istream.data =  encode_utf8(text);
```
But your solution is very handy and better

Thanks

**Reply by 117649** [reply]
When ever encounter such problem the best way is go to mozilla hg check the change sets that apply the changes. It always include some kind of solution or replacement.

**Reply by peterwx** [reply]
> When ever encounter such problem the best way is go to mozilla hg check the change sets that apply the changes. It always include some kind of solution or replacement.

Is it the best way both for scripts and legacy addons?
I ask because I mainly use scripts. 
Even without problems, one could find useful to grab functionality from mozilla code for future or existing scripts.

Change sets can span multiple files and be of a specific version update of Firefox, and there are multiple editions of Firefox.
What are more detailed steps to find solutions or functionality?

Any helpful info about this could later be included on the readme or/and with a link to a GitHub wiki by people with write access to this repo. Thanks for any provided info.

**Reply by 117649** [reply]
Use search

**Reply by peterwx** [reply]
Sure, it eluded me that searching for the specific script component erroring in the console within mozilla hg `changeset` database would be enough and straightforward, with a single query.

But then, there are tools like `mozregression` that do binary search until one finds the faulty behavior.
It seems that the latter is used mostly when there aren't errors or if they exist they're unclear, right?

For functionality, for a simple example, reading a file, one has to make an educated guess to the name of methods or modules including these and search them on mozilla hg, or by analysing scripts that fortunately make use of that functionality, no?
